### PR TITLE
FNO Inter-Foil Coupling: spectral convolution in the tandem gap region

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Every inter-foil coupling tried so far (fore-aft cross-attention, GALE, cross-DSDF) used attention — all failed. The wake is a spatially structured **frequency-domain** phenomenon: the fore foil sheds vortices at specific frequencies that interact with the aft foil's leading edge. An FNO layer in the gap region encodes this as 1D spectral convolution — the right inductive bias. NOT a full-field FNO. Operates only in the narrow coupling zone. Targets p_tan.

**References:** Li et al. "FNO for Parametric PDEs" (arXiv:2010.08895); AFNO (arXiv:2111.13587).

### Config
```python
fno_inter_foil: bool = False
fno_modes: int = 16
fno_gap_bins: int = 32
```

### FNOCouplingLayer
```python
class FNOCouplingLayer(nn.Module):
    def __init__(self, n_hidden=192, n_modes=16, G=32):
        super().__init__()
        self.n_modes = n_modes
        self.G = G
        self.fourier_weight = nn.Parameter(
            torch.randn(n_hidden, n_hidden, n_modes, dtype=torch.cfloat) * 0.02
        )
        self.bypass = nn.Linear(n_hidden, n_hidden)
        nn.init.zeros_(self.bypass.weight)
        nn.init.zeros_(self.bypass.bias)

    def forward(self, hidden, gap_node_mask, x_coords):
        if not gap_node_mask.any():
            return hidden
        B, N, C = hidden.shape
        G = self.G
        gap_hidden = hidden[:, gap_node_mask, :]
        gap_x = x_coords[:, gap_node_mask, 0]
        
        # Bin into G chord-wise cells
        x_min = gap_x.min(dim=1, keepdim=True).values
        x_max = gap_x.max(dim=1, keepdim=True).values
        x_norm = (gap_x - x_min) / (x_max - x_min + 1e-6)
        bin_idx = (x_norm * (G - 1)).long().clamp(0, G-1)
        
        # Scatter into grid
        grid = torch.zeros(B, G, C, device=hidden.device)
        counts = torch.zeros(B, G, 1, device=hidden.device)
        grid.scatter_add_(1, bin_idx.unsqueeze(-1).expand(-1,-1,C), gap_hidden)
        counts.scatter_add_(1, bin_idx.unsqueeze(-1), torch.ones_like(bin_idx.unsqueeze(-1).float()))
        grid = grid / counts.clamp(min=1)
        
        # Spectral convolution
        grid_fft = torch.fft.rfft(grid, dim=1)
        modes = min(self.n_modes, grid_fft.shape[1])
        out_fft = torch.zeros_like(grid_fft)
        out_fft[:, :modes, :] = torch.einsum('bmc,ccm->bmc',
            grid_fft[:, :modes, :].to(torch.cfloat),
            self.fourier_weight[:, :, :modes])
        grid_out = torch.fft.irfft(out_fft, n=G, dim=1)
        
        # Gather + residual
        corrections = grid_out.gather(1, bin_idx.unsqueeze(-1).expand(-1,-1,C))
        result = hidden.clone()
        result[:, gap_node_mask, :] = hidden[:, gap_node_mask, :] + self.bypass(corrections)
        return result
```

### Gap node identification
Nodes where x-coord is between fore-foil TE x and aft-foil LE x. Use existing TE coordinate features or find from surface node x-ranges. For single-foil samples: no-op (gap_node_mask is empty).

### Integration
Add between TransolverBlock 1 and 2. In `Transolver.__init__`:
```python
if cfg.fno_inter_foil:
    self.fno_coupling = FNOCouplingLayer(cfg.n_hidden, cfg.fno_modes, cfg.fno_gap_bins)
```

### Run 2 seeds
```bash
cd cfd_tandemfoil && python train.py --agent edward \
  --wandb_name "edward/fno-inter-foil-s42" \
  --wandb_group "edward/fno-inter-foil" --seed 42 \
  --fno_inter_foil --fno_modes 16 --fno_gap_bins 32 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature
# Seed 73 — same with --seed 73
```

### Report
Table: p_in, p_oodc, p_tan, p_re. W&B run IDs.

## Baseline (PR #2251)
| Metric | Value | Target |
|--------|-------|--------|
| p_in | 11.891 | < 11.89 |
| p_oodc | 7.561 | < 7.56 |
| p_tan | 28.118 | < 28.12 |
| p_re | 6.364 | < 6.36 |

W&B: 7jix2jkg (s42), epkfhxfl (s73)

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT edward:

## Results

### FNO Inter-Foil Coupling: Spectral Convolution in Tandem Gap Region

Implemented a 1D Fourier Neural Operator (FNO) coupling layer inserted between TransolverBlocks 0 and 1. The layer identifies gap-region volume nodes (between fore-foil TE and aft-foil LE), bins them into a regular 32-cell chord-wise grid, applies spectral convolution (FFT → learned weight multiply → IFFT with 16 modes), then scatters corrections back. Zero-initialized bypass ensures the layer starts as identity. Single-foil samples are no-op (empty gap mask).

### Metrics

| Metric | Baseline | Seed 42 | Seed 73 | 2-seed Avg | Δ vs Baseline |
|--------|----------|---------|---------|------------|---------------|
| p_in   | **11.891** | 14.612 | 12.314 | **13.463** | +1.572 ❌ |
| p_oodc | **7.561**  | 8.6    | 8.6    | **8.600**  | +1.039 ❌ |
| p_tan  | **28.118** | 29.7   | 29.1   | **29.400** | +1.282 ❌ |
| p_re   | **6.364**  | 6.8    | 6.8    | **6.800**  | +0.436 ❌ |

**Val loss:** s42=0.4008 (epoch 139), s73=0.3894 (epoch 140)

### Surface MAE (in-dist, from W&B)

| Component | Seed 42 | Seed 73 |
|-----------|---------|---------|
| Ux        | 2.764   | 2.722   |
| Uy        | 0.840   | 0.939   |
| p         | 14.612  | 12.314  |

### Volume MAE (in-dist, from W&B)

| Component | Seed 42 | Seed 73 |
|-----------|---------|---------|
| Ux        | 0.606   | 0.501   |
| Uy        | 0.215   | 0.196   |
| p         | 14.241  | 12.757  |

### Commands

```bash
# Seed 42
cd cfd_tandemfoil && CUDA_VISIBLE_DEVICES=0 python train.py --agent edward \
  --wandb_name "edward/fno-inter-foil-s42" --wandb_group "edward/fno-inter-foil" \
  --fno_inter_foil --fno_modes 16 --fno_gap_bins 32 \
  --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature

# Seed 73 — identical but --seed 73 --wandb_name "edward/fno-inter-foil-s73"
```

### W&B Runs
- Seed 42: `3dek688g` — https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/3dek688g
- Seed 73: `mwkdjsh7` — https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/mwkdjsh7

### Peak Memory
Both runs: ~72-83 GB on H100. No OOM issues.

### What happened

**Clear negative result — FNO coupling hurts all metrics substantially.**

- **p_in regressed by +1.572** (11.891 → 13.463). Seed 42 was particularly bad at 14.612, while seed 73 was closer to baseline at 12.314. High seed variance suggests the FNO layer introduces instability in the learning dynamics.

- **p_oodc regressed by +1.039** (7.561 → 8.600). Both seeds converged to the same 8.6, suggesting a systematic degradation rather than random variance.

- **p_tan regressed by +1.282** (28.118 → 29.400). The FNO coupling was specifically designed to help p_tan, but instead it hurt it. The spectral convolution in the gap region does not capture the inter-foil wake interaction in a way that helps the model.

- **p_re regressed by +0.436** (6.364 → 6.800). Again both seeds converged to 6.8.

### Analysis

The FNO coupling layer was a negative intervention across all metrics. Several factors likely explain this:

1. **Disruption of learned representations**: Inserting a new processing layer between blocks 0 and 1 disrupts the information flow that the baseline has optimized. Even with zero-init bypass (so the layer starts as identity), during training the spectral convolution introduces noise that the downstream blocks haven't been trained to handle.

2. **Sparse gap activation**: The FNO only fires on tandem samples (which are a minority of training data). This creates an asymmetry — the model must learn two different information paths (with and without FNO), which complicates optimization. The W&B `train/fno_gap_nodes` metric showed gap nodes only appear in tandem batches.

3. **Wrong inductive bias for this mesh**: The 1D chord-wise binning (scatter → regular grid → spectral conv → gather) loses the 2D spatial structure of the wake. The wake is a 2D phenomenon with both chord-wise and transverse components. Collapsing to 1D discards the vertical structure that matters for pressure recovery on the aft foil.

4. **Complex parameter optimization**: The `fourier_weight` is a complex-valued parameter being optimized by Lion (designed for real parameters). While PyTorch supports complex gradient updates, the optimization dynamics may be suboptimal — Lion's sign-based updates don't naturally handle the phase relationships in complex weights.

5. **Best epochs at 139-140** (vs baseline ~148): The FNO model peaked earlier and with worse loss, suggesting the extra capacity from the FNO layer isn't being utilized effectively and may be causing premature convergence.

### Suggested follow-ups
- **Abandon FNO inter-foil coupling** — the hypothesis that spectral convolution captures wake structure is not supported. Every inter-foil coupling attempt (attention, GALE, cross-DSDF, now FNO) has failed, suggesting the Transolver's existing attention mechanism may already capture inter-foil interactions implicitly.
- **If revisiting**: Try making the FNO a 2D operator (x,y grid in the gap region) rather than 1D, or apply it as a post-processing step after the last block rather than between early blocks.
- **Focus on other approaches for p_tan**: The consistent failure of explicit coupling mechanisms suggests p_tan improvements may come from better target representations, augmentation strategies, or loss reweighting rather than architectural changes to model the gap physics.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2274 because ADVISOR: Closing — all 4 metrics regressed (p_in +13.2%, p_oodc +13.7%, p_tan +4.6%, p_re +6.9%).

**Root cause:** Good analysis. This is the 5th inter-foil coupling approach to fail (fore-aft cross-attention, GALE, cross-DSDF, GNN boundary layer, FNO spectral). The pattern is now clear: **the Transolver's existing slice attention already captures inter-foil interactions implicitly**. Adding explicit coupling mechanisms — whether global (attention) or local (GNN) or spectral (FNO) — disrupts the well-tuned information flow without adding new information.

**Additional factors:** 1D chord-wise binning discards vertical wake structure. Sparse gap activation (tandem-only) creates asymmetric optimization. Complex-valued spectral weights don't play well with Lion's sign-based updates.

**Key takeaway:** ALL explicit inter-foil coupling mechanisms are dead ends. p_tan improvements must come from elsewhere — better target representations, data augmentation, or training strategies, NOT architectural coupling. Added to DO NOT REVISIT.

Moving you to a completely different direction..
